src/poetry/puzzle/solver.py

```diff
@@ -179,16 +179,14 @@ def _solve(self) -> tuple[list[Package], list[int]]:
                             if _package.name == dep.name:
                                 continue
 
-                            try:
-                                index = _package.requires.index(dep)
-                            except ValueError:
-                                _package.add_dependency(dep)
-                            else:
-                                _dep = _package.requires[index]
-                                if _dep.marker != dep.marker:
-                                    # marker of feature package is more accurate
-                                    # because it includes relevant extras
-                                    _dep.marker = dep.marker
+                            # Avoid duplication.
+                            if any(
+                                _dep == dep and _dep.marker == dep.marker
+                                for _dep in _package.requires
+                            ):
+                                continue
+
+                            _package.add_dependency(dep)
```

Comment:
```suggestion
                            try:
                                index = _package.requires.index(dep)
                            except ValueError:
                                _package.add_dependency(dep)
                            else:
                                _dep = _package.requires[index]
                                if _dep.marker != dep.marker:
                                    # merge markers
                                    _dep.marker = _dep.marker.union(dep.marker)
```

That seems to be an alternative solution that results in

```
requests = {version = "*", optional = true, markers = "extra == \"tracking-server\" or extra == \"cli\""}
```

whereas the original fix results in

```
requests = [
    {version = "*", optional = true, markers = "extra == \"cli\""},
    {version = "*", optional = true, markers = "extra == \"tracking-server\""},
]
```

Do you think one is better than the other or is it just a matter of taste?

Reply:
I spotted that possibility and decided that it didn't much matter, perhaps it is a little more cautious to preserve the separate dependencies as we find them rather than mangle them.  

I find that the way that I have written it is a little clearer, but certainly not in an important way, probably it is mostly a matter of taste.